pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -161,19 +177,103 @@ func (tb *TrustBundle) UpdateTrustAnchor(anchorConfig *TrustAnchorUpdate) error
 	return nil
 }
 
+func (tb *TrustBundle) updateRemoteEndpoint(spiffeEndpoints []string) {
+	tb.mutex.RLock()
+	remoteEndpoints := tb.endpoints
+	tb.mutex.RUnlock()
+
+	endpointChange := !isEqSpliceStr(spiffeEndpoints, remoteEndpoints)
+	if !endpointChange {
+		return
+	}
+	trustBundleLog.Infof("updated trustAnchor Remote Endpoint  :%v", spiffeEndpoints)
+
+	tb.mutex.Lock()
+	tb.endpoints = spiffeEndpoints
+	tb.endpointUpdateChan <- struct{}{}
+	tb.mutex.Unlock()
+}
+
 // AddMeshConfigUpdate : Update trustAnchor configurations from meshConfig
 func (tb *TrustBundle) AddMeshConfigUpdate(cfg *meshconfig.MeshConfig) {
+	var err error
 	if cfg != nil {
 		certs := []string{}
+		endpoints := []string{}
 		for _, pemCert := range cfg.GetCaCertificates() {
-			certs = append(certs, pemCert.GetPem())
+			cert := pemCert.GetPem()
+			if cert != "" {
+				certs = append(certs, cert)
+			} else if pemCert.GetSpiffeBundleUrl() != "" {
+				endpoints = append(endpoints, pemCert.GetSpiffeBundleUrl())
+			}
 		}
-		err := tb.UpdateTrustAnchor(&TrustAnchorUpdate{
+
+		err = tb.UpdateTrustAnchor(&TrustAnchorUpdate{
 			TrustAnchorConfig: TrustAnchorConfig{Certs: certs},
 			Source:            SourceMeshConfig,
 		})
 		if err != nil {
-			trustBundleLog.Errorf("failed to update meshConfig trustAnchors: %v", err)
+			trustBundleLog.Errorf("failed to update meshConfig PEM trustAnchors: %v", err)
+		}
+
+		tb.updateRemoteEndpoint(endpoints)
+	}
+}
+
+func (tb *TrustBundle) fetchRemoteTrustAnchors() {
+	var err error
+
+	tb.endpointMutex.RLock()
+	remoteEndpoints := tb.endpoints
+	tb.endpointMutex.RUnlock()
+	remoteCerts := []string{}
+
+	if tb.remoteCaCertPool == nil {
+		tb.remoteCaCertPool, err = x509.SystemCertPool()
+		if err != nil {
+			trustBundleLog.Error("unable to retrieve certs to validate spiffe endpoint")
+			return
+		}
+	}
+
+	currentTrustDomain := spiffe.GetTrustDomain()
+	for _, endpoint := range remoteEndpoints {
+		trustDomainAnchorMap, err := spiffe.RetrieveSpiffeBundleRootCerts(
+			map[string]string{currentTrustDomain: endpoint}, tb.remoteCaCertPool, remoteTimeout)
+		if err != nil {
+			trustBundleLog.Errorf("unable to fetch trust Anchors from endpoint %s: %s", endpoint, err)
+			continue
+		}
+		certs := trustDomainAnchorMap[currentTrustDomain]
+		for _, cert := range certs {
+			certStr := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
+			trustBundleLog.Debugf("from endpoint %v, fetched trust anchor cert: %v", endpoint, certStr)
+			remoteCerts = append(remoteCerts, certStr)
+		}
+	}
+	err = tb.UpdateTrustAnchor(&TrustAnchorUpdate{
+		TrustAnchorConfig: TrustAnchorConfig{Certs: remoteCerts},
+		Source:            sourceSpiffeEndpoints,
+	})
+	if err != nil {
+		trustBundleLog.Errorf("failed to update meshConfig Spiffe trustAnchors: %v", err)
+	}
+}
+
+func (tb *TrustBundle) ProcessRemoteTrustAnchors(stop <-chan struct{}, pollInterval time.Duration) {
```

Comment:
Why do we need this? When we get an update, just fetch the certs. we don't need to over engineer this, this code is triggered ~once a year most likely

Reply:
And it doesn't even matter if it gets blocked for a minute if we had a ton of certs and slow endpoints - no one needs rapid root cert transitions

Reply:
I am fine with fetching trustAnchors only at time of configuration. However,  @myidpt - can you weigh in on this? I thought we had decided that *some* form of polling of the spiffe bundle endpoint was always going to be required - I have set it for every half hour, but this could be once a day or even once a month. 

Once we start supporting spiffe endpoints per trustdomain, do we not have to poll it periodically?

Reply:
Oh I actually read this wrong, I thought this was to make the updates async - didn't realize it was polling.

This already has expected behavior defined in the spec: https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#412-refresh-hint

Since its optional, I recommend checking existing implementations (spire?) and copying them. or if there is a common spiffe library to do it - even better

Reply:
The go-routine was mainly added to periodically poll the endpoints. Even if polling is done just once a year, we need to have some routine to do this periodically. If we decide to do away with polling periodically, can remove this code altogether.

Reply:
Went through some code [here](https://github.com/spiffe/spire/blob/4387f6b0cf1e95fdd25d25606feb38a3844ec351/cmd/spire-agent/cli/api/fetch_x509.go#L183). It does appear that fetching the trustbundle and the spiffe cert happens concurrently on the agent. Also, as per standards [here](https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#412-refresh-hint), the client need not adhere to the refresh-hint provided. Perhaps this ability to honor the refresh-hint could be added in a subsequent PR?

Reply:
@myidpt @costinm any experience with this?

Reply:
I think we should periodically fetch the root. One reason for that is users don't have control over the control planes when it's managed. The default frequency should be very low - once every week seems reasonable to me. It's OK to leave a TODO to respect the spiffe_refresh_hint. But that's a low priority - root transition typically happen in the scale quarters or beyond. If there's a hard transitioning (due to root compromise), customers should take more explicit actions including restarting all affected control planes.